In [1]:
# 성능 평가 계산 함수
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.metrics import precision_score, recall_score
from sklearn.metrics import f1_score

def get_clf_eval(y_test, pred):
    accuracy = accuracy_score(y_test, pred)
    precision = precision_score(y_test, pred, average='macro')
    recall = recall_score(y_test, pred, average='macro')
    f1 = f1_score(y_test, pred, average='macro')
    
    print('정확도: {0:.4f}, 정밀도: {1:.4f}, 재현율: {2:.4f}, F1:{3:.4f}'.format(accuracy, precision, recall, f1))

# 의사 결정 트리 (DecisionTree Classifier)

In [2]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import GridSearchCV

# DecisionTree Classifier 생성
dt_clf = DecisionTreeClassifier(random_state=0)

# MNIST 데이터 로딩
digits = load_digits()

# 학습 데이터와 테스트 데이터로 분리
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target, test_size=0.2, random_state=0)

# parameter들을 dictionary 형태로 설정
parameters = {'max_depth':[2,3,4,5], 'min_samples_split':[2,4,6,8,10]}

#### 정확도

In [3]:
# 교차 검증 및 하이퍼 파라미터 최적화
import pandas as pd

# param_grid의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트 수행 설정
cross_validation = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
grid_dtree = GridSearchCV(dt_clf, param_grid=parameters, cv=cross_validation, refit=True)

# MNIST 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습 및 평가
grid_dtree.fit(X_train, y_train)

# GridSearchCV 결과 추출하여 DataFrame으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', \
           'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score']]

params  mean_test_score  \
0    {'max_depth': 2, 'min_samples_split': 2}         0.329152   
1    {'max_depth': 2, 'min_samples_split': 4}         0.329152   
2    {'max_depth': 2, 'min_samples_split': 6}         0.329152   
3    {'max_depth': 2, 'min_samples_split': 8}         0.329152   
4   {'max_depth': 2, 'min_samples_split': 10}         0.329152   
5    {'max_depth': 3, 'min_samples_split': 2}         0.478777   
6    {'max_depth': 3, 'min_samples_split': 4}         0.478777   
7    {'max_depth': 3, 'min_samples_split': 6}         0.478777   
8    {'max_depth': 3, 'min_samples_split': 8}         0.478777   
9   {'max_depth': 3, 'min_samples_split': 10}         0.478777   
10   {'max_depth': 4, 'min_samples_split': 2}         0.677108   
11   {'max_depth': 4, 'min_samples_split': 4}         0.677108   
12   {'max_depth': 4, 'min_samples_split': 6}         0.677108   
13   {'max_depth': 4, 'min_samples_split': 8}         0.677108   
14  {'max_depth': 4, 'min_samples_split': 10}         0.676413   
15   {'max_depth': 5, 'min_samples_split': 2}         0.748754   
16   {'max_depth': 5, 'min_samples_split': 4}         0.748754   
17   {'max_depth': 5, 'min_samples_split': 6}         0.748754   
18   {'max_depth': 5, 'min_samples_split': 8}         0.748059   
19  {'max_depth': 5, 'min_samples_split': 10}         0.747363   

    rank_test_score  split0_test_score  split1_test_score  split2_test_score  \
0                16           0.333333           0.333333           0.324042   
1                16           0.333333           0.333333           0.324042   
2                16           0.333333           0.333333           0.324042   
3                16           0.333333           0.333333           0.324042   
4                16           0.333333           0.333333           0.324042   
5                11           0.461806           0.493056           0.494774   
6                11           0.461806           0.493056           0.494774   
7                11           0.461806           0.493056           0.494774   
8                11           0.461806           0.493056           0.494774   
9                11           0.461806           0.493056           0.494774   
10                6           0.666667           0.684028           0.707317   
11                6           0.666667           0.684028           0.707317   
12                6           0.666667           0.684028           0.707317   
13                6           0.666667           0.684028           0.707317   
14               10           0.663194           0.684028           0.707317   
15                1           0.781250           0.756944           0.742160   
16                1           0.781250           0.756944           0.742160   
17                1           0.781250           0.756944           0.742160   
18                4           0.777778           0.756944           0.742160   
19                5           0.777778           0.756944           0.742160   

    split3_test_score  split4_test_score  
0            0.320557           0.334495  
1            0.320557           0.334495  
2            0.320557           0.334495  
3            0.320557           0.334495  
4            0.320557           0.334495  
5            0.477352           0.466899  
6            0.477352           0.466899  
7            0.477352           0.466899  
8            0.477352           0.466899  
9            0.477352           0.466899  
10           0.651568           0.675958  
11           0.651568           0.675958  
12           0.651568           0.675958  
13           0.651568           0.675958  
14           0.651568           0.675958  
15           0.717770           0.745645  
16           0.717770           0.745645  
17           0.717770           0.745645  
18           0.717770           0.745645  
19           0.717770           0.742160

In [4]:
print('GridSearchCV 최적 파라미터:',grid_dtree.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_dtree.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 5, 'min_samples_split': 2}
GridSearchCV 최고 정확도: 0.7488


=> DecisionTree Classifier 모델에서 하이퍼 파라미터 최적화 결과 max_depth가 5, min_samples_split이 2일 때 평균 최고 정확도가 74.88%로
측정됐다.

#### 정밀도

In [5]:
# 교차 검증 및 하이퍼 파라미터 최적화
import pandas as pd
from sklearn.metrics import make_scorer
import warnings
warnings.filterwarnings('ignore')

# param_grid의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트 수행 설정
cross_validation = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
grid_dtree = GridSearchCV(dt_clf, param_grid=parameters, scoring=make_scorer(precision_score, average='macro'), cv=cross_validation, refit=True)

# MNIST 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습 및 평가
grid_dtree.fit(X_train, y_train)

# GridSearchCV 결과 추출하여 DataFrame으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', \
           'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score']]

params  mean_test_score  \
0    {'max_depth': 2, 'min_samples_split': 2}         0.203680   
1    {'max_depth': 2, 'min_samples_split': 4}         0.203680   
2    {'max_depth': 2, 'min_samples_split': 6}         0.203680   
3    {'max_depth': 2, 'min_samples_split': 8}         0.203680   
4   {'max_depth': 2, 'min_samples_split': 10}         0.203680   
5    {'max_depth': 3, 'min_samples_split': 2}         0.443988   
6    {'max_depth': 3, 'min_samples_split': 4}         0.443988   
7    {'max_depth': 3, 'min_samples_split': 6}         0.443988   
8    {'max_depth': 3, 'min_samples_split': 8}         0.443988   
9   {'max_depth': 3, 'min_samples_split': 10}         0.443988   
10   {'max_depth': 4, 'min_samples_split': 2}         0.605805   
11   {'max_depth': 4, 'min_samples_split': 4}         0.605805   
12   {'max_depth': 4, 'min_samples_split': 6}         0.605805   
13   {'max_depth': 4, 'min_samples_split': 8}         0.605805   
14  {'max_depth': 4, 'min_samples_split': 10}         0.605274   
15   {'max_depth': 5, 'min_samples_split': 2}         0.749863   
16   {'max_depth': 5, 'min_samples_split': 4}         0.749610   
17   {'max_depth': 5, 'min_samples_split': 6}         0.749605   
18   {'max_depth': 5, 'min_samples_split': 8}         0.749091   
19  {'max_depth': 5, 'min_samples_split': 10}         0.749332   

    rank_test_score  split0_test_score  split1_test_score  split2_test_score  \
0                16           0.221763           0.198119           0.207013   
1                16           0.221763           0.198119           0.207013   
2                16           0.221763           0.198119           0.207013   
3                16           0.221763           0.198119           0.207013   
4                16           0.221763           0.198119           0.207013   
5                11           0.383077           0.457094           0.467531   
6                11           0.383077           0.457094           0.467531   
7                11           0.383077           0.457094           0.467531   
8                11           0.383077           0.457094           0.467531   
9                11           0.383077           0.457094           0.467531   
10                6           0.583117           0.651731           0.647208   
11                6           0.583117           0.651731           0.647208   
12                6           0.583117           0.651731           0.647208   
13                6           0.583117           0.651731           0.647208   
14               10           0.580466           0.651731           0.647208   
15                1           0.742205           0.781656           0.779271   
16                2           0.742205           0.778701           0.779271   
17                3           0.742205           0.778701           0.779247   
18                5           0.739447           0.778891           0.779247   
19                4           0.743024           0.778891           0.779247   

    split3_test_score  split4_test_score  
0            0.193911           0.197595  
1            0.193911           0.197595  
2            0.193911           0.197595  
3            0.193911           0.197595  
4            0.193911           0.197595  
5            0.463977           0.448261  
6            0.463977           0.448261  
7            0.463977           0.448261  
8            0.463977           0.448261  
9            0.463977           0.448261  
10           0.564884           0.582083  
11           0.564884           0.582083  
12           0.564884           0.582083  
13           0.564884           0.582083  
14           0.564884           0.582083  
15           0.727646           0.718536  
16           0.729335           0.718536  
17           0.729335           0.718536  
18           0.729335           0.718536  
19           0.729335           0.716165

In [6]:
print('GridSearchCV 최적 파라미터:',grid_dtree.best_params_)
print('GridSearchCV 최고 정밀도: {0:.4f}'.format(grid_dtree.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 5, 'min_samples_split': 2}
GridSearchCV 최고 정밀도: 0.7499


=> DecisionTree Classifier 모델에서 하이퍼 파라미터 최적화 결과 max_depth가 5, min_samples_split이 2일 때 평균 최고 정밀도가 74.99%로
측정됐다

#### 재현율

In [7]:
# 교차 검증 및 하이퍼 파라미터 최적화
import pandas as pd
from sklearn.metrics import make_scorer
import warnings
warnings.filterwarnings('ignore')

# param_grid의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트 수행 설정
cross_validation = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
grid_dtree = GridSearchCV(dt_clf, param_grid=parameters, scoring=make_scorer(recall_score, average='macro'), cv=cross_validation, refit=True)

# MNIST 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습 및 평가
grid_dtree.fit(X_train, y_train)

# GridSearchCV 결과 추출하여 DataFrame으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', \
           'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score']]

params  mean_test_score  \
0    {'max_depth': 2, 'min_samples_split': 2}         0.313750   
1    {'max_depth': 2, 'min_samples_split': 4}         0.313750   
2    {'max_depth': 2, 'min_samples_split': 6}         0.313750   
3    {'max_depth': 2, 'min_samples_split': 8}         0.313750   
4   {'max_depth': 2, 'min_samples_split': 10}         0.313750   
5    {'max_depth': 3, 'min_samples_split': 2}         0.470694   
6    {'max_depth': 3, 'min_samples_split': 4}         0.470694   
7    {'max_depth': 3, 'min_samples_split': 6}         0.470694   
8    {'max_depth': 3, 'min_samples_split': 8}         0.470694   
9   {'max_depth': 3, 'min_samples_split': 10}         0.470694   
10   {'max_depth': 4, 'min_samples_split': 2}         0.673970   
11   {'max_depth': 4, 'min_samples_split': 4}         0.673970   
12   {'max_depth': 4, 'min_samples_split': 6}         0.673970   
13   {'max_depth': 4, 'min_samples_split': 8}         0.673970   
14  {'max_depth': 4, 'min_samples_split': 10}         0.673255   
15   {'max_depth': 5, 'min_samples_split': 2}         0.744620   
16   {'max_depth': 5, 'min_samples_split': 4}         0.744620   
17   {'max_depth': 5, 'min_samples_split': 6}         0.744620   
18   {'max_depth': 5, 'min_samples_split': 8}         0.743905   
19  {'max_depth': 5, 'min_samples_split': 10}         0.743143   

    rank_test_score  split0_test_score  split1_test_score  split2_test_score  \
0                16           0.318810           0.318810           0.310038   
1                16           0.318810           0.318810           0.310038   
2                16           0.318810           0.318810           0.310038   
3                16           0.318810           0.318810           0.310038   
4                16           0.318810           0.318810           0.310038   
5                11           0.452834           0.484631           0.489506   
6                11           0.452834           0.484631           0.489506   
7                11           0.452834           0.484631           0.489506   
8                11           0.452834           0.484631           0.489506   
9                11           0.452834           0.484631           0.489506   
10                6           0.664132           0.682765           0.704968   
11                6           0.664132           0.682765           0.704968   
12                6           0.664132           0.682765           0.704968   
13                6           0.664132           0.682765           0.704968   
14               10           0.660561           0.682765           0.704968   
15                1           0.774954           0.755756           0.740525   
16                1           0.774954           0.755756           0.740525   
17                1           0.774954           0.755756           0.740525   
18                4           0.771382           0.755756           0.740525   
19                5           0.771144           0.755756           0.740525   

    split3_test_score  split4_test_score  
0            0.305323           0.315771  
1            0.305323           0.315771  
2            0.305323           0.315771  
3            0.305323           0.315771  
4            0.305323           0.315771  
5            0.470568           0.455929  
6            0.470568           0.455929  
7            0.470568           0.455929  
8            0.470568           0.455929  
9            0.470568           0.455929  
10           0.648864           0.669120  
11           0.648864           0.669120  
12           0.648864           0.669120  
13           0.648864           0.669120  
14           0.648864           0.669120  
15           0.713173           0.738689  
16           0.713173           0.738689  
17           0.713173           0.738689  
18           0.713173           0.738689  
19           0.713173           0.735118

In [8]:
print('GridSearchCV 최적 파라미터:',grid_dtree.best_params_)
print('GridSearchCV 최고 재현율: {0:.4f}'.format(grid_dtree.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 5, 'min_samples_split': 2}
GridSearchCV 최고 재현율: 0.7446


=> DecisionTree Classifier 모델에서 하이퍼 파라미터 최적화 결과 max_depth가 5, min_samples_split이 2일 때 평균 최고 재현율이 74.46%로
측정됐다.

#### F1 스코어

In [9]:
# 교차 검증 및 하이퍼 파라미터 최적화
import pandas as pd
from sklearn.metrics import make_scorer
import warnings
warnings.filterwarnings('ignore')

# param_grid의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트 수행 설정
cross_validation = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
grid_dtree = GridSearchCV(dt_clf, param_grid=parameters, scoring=make_scorer(f1_score, average='macro'), cv=cross_validation, refit=True)

# MNIST 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습 및 평가
grid_dtree.fit(X_train, y_train)

# GridSearchCV 결과 추출하여 DataFrame으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', \
           'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score']]

params  mean_test_score  \
0    {'max_depth': 2, 'min_samples_split': 2}         0.209319   
1    {'max_depth': 2, 'min_samples_split': 4}         0.209319   
2    {'max_depth': 2, 'min_samples_split': 6}         0.209319   
3    {'max_depth': 2, 'min_samples_split': 8}         0.209319   
4   {'max_depth': 2, 'min_samples_split': 10}         0.209319   
5    {'max_depth': 3, 'min_samples_split': 2}         0.383898   
6    {'max_depth': 3, 'min_samples_split': 4}         0.383898   
7    {'max_depth': 3, 'min_samples_split': 6}         0.383898   
8    {'max_depth': 3, 'min_samples_split': 8}         0.383898   
9   {'max_depth': 3, 'min_samples_split': 10}         0.383898   
10   {'max_depth': 4, 'min_samples_split': 2}         0.618221   
11   {'max_depth': 4, 'min_samples_split': 4}         0.618221   
12   {'max_depth': 4, 'min_samples_split': 6}         0.618221   
13   {'max_depth': 4, 'min_samples_split': 8}         0.618221   
14  {'max_depth': 4, 'min_samples_split': 10}         0.617591   
15   {'max_depth': 5, 'min_samples_split': 2}         0.734098   
16   {'max_depth': 5, 'min_samples_split': 4}         0.733909   
17   {'max_depth': 5, 'min_samples_split': 6}         0.733895   
18   {'max_depth': 5, 'min_samples_split': 8}         0.733194   
19  {'max_depth': 5, 'min_samples_split': 10}         0.732899   

    rank_test_score  split0_test_score  split1_test_score  split2_test_score  \
0                16           0.213329           0.209566           0.215841   
1                16           0.213329           0.209566           0.215841   
2                16           0.213329           0.209566           0.215841   
3                16           0.213329           0.209566           0.215841   
4                16           0.213329           0.209566           0.215841   
5                11           0.367253           0.390396           0.394844   
6                11           0.367253           0.390396           0.394844   
7                11           0.367253           0.390396           0.394844   
8                11           0.367253           0.390396           0.394844   
9                11           0.367253           0.390396           0.394844   
10                6           0.605003           0.632479           0.650767   
11                6           0.605003           0.632479           0.650767   
12                6           0.605003           0.632479           0.650767   
13                6           0.605003           0.632479           0.650767   
14               10           0.601855           0.632479           0.650767   
15                1           0.746499           0.760748           0.748052   
16                2           0.746499           0.760017           0.748052   
17                3           0.746499           0.760017           0.747986   
18                4           0.742935           0.760073           0.747986   
19                5           0.743790           0.760073           0.747986   

    split3_test_score  split4_test_score  
0            0.202254           0.205608  
1            0.202254           0.205608  
2            0.202254           0.205608  
3            0.202254           0.205608  
4            0.202254           0.205608  
5            0.390192           0.376807  
6            0.390192           0.376807  
7            0.390192           0.376807  
8            0.390192           0.376807  
9            0.390192           0.376807  
10           0.591086           0.611768  
11           0.591086           0.611768  
12           0.591086           0.611768  
13           0.591086           0.611768  
14           0.591086           0.611768  
15           0.702258           0.712931  
16           0.702045           0.712931  
17           0.702045           0.712931  
18           0.702045           0.712931  
19           0.702045           0.710601

In [10]:
print('GridSearchCV 최적 파라미터:',grid_dtree.best_params_)
print('GridSearchCV 최고 F1 스코어: {0:.4f}'.format(grid_dtree.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 5, 'min_samples_split': 2}
GridSearchCV 최고 F1 스코어: 0.7341


=> DecisionTree Classifier 모델에서 하이퍼 파라미터 최적화 결과 max_depth가 5, min_samples_split이 2일 때 평균 최고 F1 스코어가 73.41%로
측정됐다.

In [11]:
# GridSearchCV의 refit으로 이미 학습된 estimator 반환
estimator = grid_dtree.best_estimator_

# 테스트 데이터 세트의 성능 평가
pred = estimator.predict(X_test)
get_clf_eval(y_test,pred)

정확도: 0.7306, 정밀도: 0.7656, 재현율: 0.7368, F1:0.7409


=> 별도의 테스트 데이터 세트로 성능을 평가한 결과 정확도는 73.06%, 정밀도는 76.56%, 재현율은 73.68%, 그리고 F1 스코어는 74.09%의 결과가 도출됐다.
=> 위의 성능 평가를 종합한 결과 max_depth가 5, min_samples_split이 2일 때 DecisionTree Classifier 모델은 가장 최적의 성능을 보인다.

# 랜덤 포레스트 (Random Forest Classifier)

In [12]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import GridSearchCV

# Random Forest Classifier 생성
rf_clf = RandomForestClassifier(random_state=0)

# MNIST 데이터 로딩
digits = load_digits()

# 학습 데이터와 테스트 데이터로 분리
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target, test_size=0.2, random_state=0)

# parameter들을 dictionary 형태로 설정
parameters = {'min_samples_split':[2,4,6,8,10], 'n_estimators':[100, 300, 500]} 

#### 정확도

In [13]:
# 하이퍼 파라미터 최적화 및 교차검증
import pandas as pd

cross_validation = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
grid_rforest = GridSearchCV(rf_clf, param_grid=parameters, cv=cross_validation, refit=True)

grid_rforest.fit(X_train, y_train)

# GridSearchCV 결과 추출하여 DataFrame으로 변환
scores_df = pd.DataFrame(grid_rforest.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', \
           'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score']]

params  mean_test_score  \
0    {'min_samples_split': 2, 'n_estimators': 100}         0.972849   
1    {'min_samples_split': 2, 'n_estimators': 300}         0.971462   
2    {'min_samples_split': 2, 'n_estimators': 500}         0.971465   
3    {'min_samples_split': 4, 'n_estimators': 100}         0.966592   
4    {'min_samples_split': 4, 'n_estimators': 300}         0.966589   
5    {'min_samples_split': 4, 'n_estimators': 500}         0.968682   
6    {'min_samples_split': 6, 'n_estimators': 100}         0.967289   
7    {'min_samples_split': 6, 'n_estimators': 300}         0.970074   
8    {'min_samples_split': 6, 'n_estimators': 500}         0.969377   
9    {'min_samples_split': 8, 'n_estimators': 100}         0.965203   
10   {'min_samples_split': 8, 'n_estimators': 300}         0.966594   
11   {'min_samples_split': 8, 'n_estimators': 500}         0.967289   
12  {'min_samples_split': 10, 'n_estimators': 100}         0.963117   
13  {'min_samples_split': 10, 'n_estimators': 300}         0.963117   
14  {'min_samples_split': 10, 'n_estimators': 500}         0.965200   

    rank_test_score  split0_test_score  split1_test_score  split2_test_score  \
0                 1           0.986111           0.975694           0.961672   
1                 3           0.986111           0.965278           0.961672   
2                 2           0.982639           0.965278           0.968641   
3                10           0.982639           0.958333           0.965157   
4                11           0.982639           0.961806           0.958188   
5                 6           0.982639           0.958333           0.965157   
6                 8           0.975694           0.965278           0.961672   
7                 4           0.982639           0.961806           0.968641   
8                 5           0.982639           0.961806           0.968641   
9                12           0.982639           0.951389           0.968641   
10                9           0.982639           0.954861           0.961672   
11                7           0.982639           0.958333           0.961672   
12               14           0.979167           0.947917           0.961672   
13               14           0.975694           0.951389           0.968641   
14               13           0.979167           0.958333           0.968641   

    split3_test_score  split4_test_score  
0            0.972125           0.968641  
1            0.975610           0.968641  
2            0.972125           0.968641  
3            0.965157           0.961672  
4            0.968641           0.961672  
5            0.968641           0.968641  
6            0.965157           0.968641  
7            0.968641           0.968641  
8            0.968641           0.965157  
9            0.965157           0.958188  
10           0.972125           0.961672  
11           0.972125           0.961672  
12           0.968641           0.958188  
13           0.968641           0.951220  
14           0.965157           0.954704

In [14]:
print('GridSearchCV 최적 파라미터:',grid_rforest.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_rforest.best_score_))

GridSearchCV 최적 파라미터: {'min_samples_split': 2, 'n_estimators': 100}
GridSearchCV 최고 정확도: 0.9728


=> Random Forest Classifier 모델에서 하이퍼 파라미터 최적화 결과 n_estimators가 100, min_samples_split이 2일 때 평균 최고 정확도가 97.28%로
측정됐다.

#### 정밀도

In [15]:
# 하이퍼 파라미터 최적화 및 교차검증
import pandas as pd

cross_validation = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
grid_rforest = GridSearchCV(rf_clf, param_grid=parameters,  scoring=make_scorer(precision_score, average='macro'), cv=cross_validation, refit=True)

grid_rforest.fit(X_train, y_train)

# GridSearchCV 결과 추출하여 DataFrame으로 변환
scores_df = pd.DataFrame(grid_rforest.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', \
           'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score']]

params  mean_test_score  \
0    {'min_samples_split': 2, 'n_estimators': 100}         0.973813   
1    {'min_samples_split': 2, 'n_estimators': 300}         0.972462   
2    {'min_samples_split': 2, 'n_estimators': 500}         0.972310   
3    {'min_samples_split': 4, 'n_estimators': 100}         0.967732   
4    {'min_samples_split': 4, 'n_estimators': 300}         0.967739   
5    {'min_samples_split': 4, 'n_estimators': 500}         0.969643   
6    {'min_samples_split': 6, 'n_estimators': 100}         0.968248   
7    {'min_samples_split': 6, 'n_estimators': 300}         0.970778   
8    {'min_samples_split': 6, 'n_estimators': 500}         0.970298   
9    {'min_samples_split': 8, 'n_estimators': 100}         0.965828   
10   {'min_samples_split': 8, 'n_estimators': 300}         0.967314   
11   {'min_samples_split': 8, 'n_estimators': 500}         0.968174   
12  {'min_samples_split': 10, 'n_estimators': 100}         0.964435   
13  {'min_samples_split': 10, 'n_estimators': 300}         0.963595   
14  {'min_samples_split': 10, 'n_estimators': 500}         0.966118   

    rank_test_score  split0_test_score  split1_test_score  split2_test_score  \
0                 1           0.985908           0.975757           0.963755   
1                 2           0.986273           0.966819           0.963243   
2                 3           0.983148           0.966328           0.969804   
3                10           0.982570           0.960617           0.966947   
4                 9           0.982570           0.963115           0.960818   
5                 6           0.982570           0.959968           0.966815   
6                 7           0.975418           0.966623           0.962538   
7                 4           0.982570           0.962731           0.970050   
8                 5           0.982570           0.963672           0.970050   
9                13           0.982683           0.952477           0.970050   
10               11           0.982570           0.955991           0.963243   
11                8           0.982570           0.959968           0.963243   
12               14           0.979504           0.949291           0.963958   
13               15           0.975854           0.951059           0.970050   
14               12           0.979445           0.959968           0.970050   

    split3_test_score  split4_test_score  
0            0.974267           0.969375  
1            0.977089           0.968883  
2            0.973386           0.968883  
3            0.965876           0.962653  
4            0.970563           0.961630  
5            0.969978           0.968883  
6            0.965876           0.970785  
7            0.969467           0.969071  
8            0.969467           0.965732  
9            0.965661           0.958267  
10           0.972908           0.961860  
11           0.972908           0.962182  
12           0.969674           0.959747  
13           0.969674           0.951336  
14           0.966241           0.954887

In [16]:
print('GridSearchCV 최적 파라미터:',grid_rforest.best_params_)
print('GridSearchCV 최고 정밀도: {0:.4f}'.format(grid_rforest.best_score_))

GridSearchCV 최적 파라미터: {'min_samples_split': 2, 'n_estimators': 100}
GridSearchCV 최고 정밀도: 0.9738


=> Random Forest Classifier 모델에서 하이퍼 파라미터 최적화 결과 n_estimators가 100, min_samples_split이 2일 때 평균 최고 정밀도가 97.38%로
측정됐다.

#### 재현율

In [17]:
# 하이퍼 파라미터 최적화 및 교차검증
import pandas as pd

cross_validation = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
grid_rforest = GridSearchCV(rf_clf, param_grid=parameters,  scoring=make_scorer(recall_score, average='macro'), cv=cross_validation, refit=True)

grid_rforest.fit(X_train, y_train)

# GridSearchCV 결과 추출하여 DataFrame으로 변환
scores_df = pd.DataFrame(grid_rforest.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', \
           'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score']]

params  mean_test_score  \
0    {'min_samples_split': 2, 'n_estimators': 100}         0.972493   
1    {'min_samples_split': 2, 'n_estimators': 300}         0.971182   
2    {'min_samples_split': 2, 'n_estimators': 500}         0.971108   
3    {'min_samples_split': 4, 'n_estimators': 100}         0.966202   
4    {'min_samples_split': 4, 'n_estimators': 300}         0.966271   
5    {'min_samples_split': 4, 'n_estimators': 500}         0.968302   
6    {'min_samples_split': 6, 'n_estimators': 100}         0.966888   
7    {'min_samples_split': 6, 'n_estimators': 300}         0.969736   
8    {'min_samples_split': 6, 'n_estimators': 500}         0.969020   
9    {'min_samples_split': 8, 'n_estimators': 100}         0.964651   
10   {'min_samples_split': 8, 'n_estimators': 300}         0.966271   
11   {'min_samples_split': 8, 'n_estimators': 500}         0.966967   
12  {'min_samples_split': 10, 'n_estimators': 100}         0.962546   
13  {'min_samples_split': 10, 'n_estimators': 300}         0.962502   
14  {'min_samples_split': 10, 'n_estimators': 500}         0.964724   

    rank_test_score  split0_test_score  split1_test_score  split2_test_score  \
0                 1           0.985450           0.974471           0.962548   
1                 2           0.985820           0.964340           0.962786   
2                 3           0.982249           0.963970           0.969691   
3                11           0.982249           0.957065           0.966012   
4                 9           0.982249           0.960637           0.959215   
5                 6           0.982249           0.957065           0.966012   
6                 8           0.975319           0.964102           0.962416   
7                 4           0.982249           0.960769           0.969345   
8                 5           0.982249           0.960769           0.969345   
9                13           0.982116           0.950506           0.969345   
10                9           0.982249           0.953839           0.962786   
11                7           0.982249           0.957065           0.962786   
12               14           0.978439           0.946910           0.962786   
13               15           0.974974           0.950136           0.969345   
14               12           0.978677           0.957065           0.969345   

    split3_test_score  split4_test_score  
0            0.972463           0.967532  
1            0.975911           0.967054  
2            0.972578           0.967054  
3            0.965426           0.960257  
4            0.969129           0.960125  
5            0.969129           0.967054  
6            0.965426           0.967178  
7            0.969129           0.967187  
8            0.969129           0.963606  
9            0.965080           0.956208  
10           0.972355           0.960125  
11           0.972355           0.960380  
12           0.968652           0.955943  
13           0.968652           0.949401  
14           0.965558           0.952973

In [18]:
print('GridSearchCV 최적 파라미터:',grid_rforest.best_params_)
print('GridSearchCV 최고 재현율: {0:.4f}'.format(grid_rforest.best_score_))

GridSearchCV 최적 파라미터: {'min_samples_split': 2, 'n_estimators': 100}
GridSearchCV 최고 재현율: 0.9725


=> Random Forest Classifier 모델에서 하이퍼 파라미터 최적화 결과 n_estimators가 100, min_samples_split이 2일 때 평균 최고 재현율이 97.25%로
측정됐다.

#### F1 스코어

In [19]:
# 하이퍼 파라미터 최적화 및 교차검증
import pandas as pd

cross_validation = StratifiedKFold(n_splits=5, shuffle=True, random_state=0)
grid_rforest = GridSearchCV(rf_clf, param_grid=parameters,  scoring=make_scorer(f1_score, average='macro'), cv=cross_validation, refit=True)

grid_rforest.fit(X_train, y_train)

# GridSearchCV 결과 추출하여 DataFrame으로 변환
scores_df = pd.DataFrame(grid_rforest.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', \
           'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score']]

params  mean_test_score  \
0    {'min_samples_split': 2, 'n_estimators': 100}         0.972446   
1    {'min_samples_split': 2, 'n_estimators': 300}         0.971087   
2    {'min_samples_split': 2, 'n_estimators': 500}         0.970971   
3    {'min_samples_split': 4, 'n_estimators': 100}         0.966077   
4    {'min_samples_split': 4, 'n_estimators': 300}         0.966101   
5    {'min_samples_split': 4, 'n_estimators': 500}         0.968189   
6    {'min_samples_split': 6, 'n_estimators': 100}         0.966777   
7    {'min_samples_split': 6, 'n_estimators': 300}         0.969650   
8    {'min_samples_split': 6, 'n_estimators': 500}         0.968987   
9    {'min_samples_split': 8, 'n_estimators': 100}         0.964478   
10   {'min_samples_split': 8, 'n_estimators': 300}         0.966034   
11   {'min_samples_split': 8, 'n_estimators': 500}         0.966809   
12  {'min_samples_split': 10, 'n_estimators': 100}         0.962438   
13  {'min_samples_split': 10, 'n_estimators': 300}         0.962345   
14  {'min_samples_split': 10, 'n_estimators': 500}         0.964686   

    rank_test_score  split0_test_score  split1_test_score  split2_test_score  \
0                 1           0.985556           0.974549           0.961666   
1                 2           0.985933           0.964465           0.961673   
2                 3           0.982528           0.963839           0.968634   
3                10           0.982297           0.957030           0.965299   
4                 9           0.982297           0.960510           0.958346   
5                 6           0.982297           0.957107           0.965309   
6                 8           0.975199           0.964208           0.961423   
7                 4           0.982297           0.960844           0.968686   
8                 5           0.982297           0.961062           0.968686   
9                13           0.982222           0.950045           0.968686   
10               11           0.982297           0.953730           0.961673   
11                7           0.982297           0.957107           0.961673   
12               14           0.978686           0.946537           0.961795   
13               15           0.975180           0.949515           0.968686   
14               12           0.978891           0.957107           0.968686   

    split3_test_score  split4_test_score  
0            0.972509           0.967949  
1            0.975974           0.967392  
2            0.972463           0.967392  
3            0.965075           0.960686  
4            0.968998           0.960356  
5            0.968843           0.967392  
6            0.965075           0.967981  
7            0.968786           0.967637  
8            0.968786           0.964103  
9            0.964956           0.956481  
10           0.972117           0.960356  
11           0.972117           0.960851  
12           0.968556           0.956617  
13           0.968556           0.949788  
14           0.965396           0.953348

In [20]:
print('GridSearchCV 최적 파라미터:',grid_rforest.best_params_)
print('GridSearchCV 최고 F1 스코어: {0:.4f}'.format(grid_rforest.best_score_))

GridSearchCV 최적 파라미터: {'min_samples_split': 2, 'n_estimators': 100}
GridSearchCV 최고 F1 스코어: 0.9724


=> Random Forest Classifier 모델에서 하이퍼 파라미터 최적화 결과 n_estimators가 100, min_samples_split이 2일 때 평균 최고 F1 스코어가 97.24%로
측정됐다.

In [21]:
estimator = grid_rforest.best_estimator_

pred = estimator.predict(X_test)

get_clf_eval(y_test,pred)

정확도: 0.9694, 정밀도: 0.9695, 재현율: 0.9709, F1:0.9698


=> 별도의 테스트 데이터 세트로 성능을 평가한 결과 정확도는 96.94%, 정밀도는 96.95%, 재현율은 97.09%, 그리고 F1 스코어는 96.98%의 결과가 도출됐다.
=> 위의 성능 평가를 종합한 결과 n_estimators가 100, min_samples_split이 2일 때 Random Forest Classifier 모델은 가장 최적의 성능을 보인다.

# 성능 레포트

의사 결정 트리(DecisionTree Classifier) 모델에서 하이퍼 파라미터 최적화를 수행하였을 때 정확도, 정밀도, 재현율, F1 스코어 각 성능 모두 max_depth가 5, min_samples_split이 2일 때 가장 최적 성능을 보였다. 각 성능은 정확도는 74.88%, 정밀도는 74.99%, 재현율은 74.46%, F1 스코어는 73.41%로 측정됐다. 가장 최적 성능을 보인 모델을 테스트 데이터 세트로 성능 평가를 한 결과 정확도는 73.06%, 정밀도는 76.56%, 재현율은 73.68%, 그리고 F1 스코어는 74.09%의 결과가 도출됐다.

랜덤 포레스트(Random Forest Classifier)모델에서 하이퍼 파라미터 최적화를 수행하였을 때 정확도, 정밀도, 재현율, F1 스코어 각 성능 모두 n_estimator가 100, min_samples_split이 2일 때 가장 최적 성능을 보였다. 각 성능은 정확도는 97.28%, 정밀도는 97.38%, 재현율은 97.25%, F1 스코어는 97.24%로 측정됐다. 가장 최적 성능을 보인 모델을 테스트 데이터 세트로 성능 평가를 한 결과 정확도는 96.94%, 정밀도는 96.95%, 재현율은 97.09%, 그리고 F1 스코어는 96.98%의 결과가 도출됐다.

의사 결정 트리(DecisionTree Classifier)와 랜덤 포레스트(Random Forest Classifier) 각 모델의 최적 성능을 보일 때의 두 모델을 비교하였을 때 정확도, 정밀도, 재현율, F1 스코어 각 성능 모두 랜덤 포레스트가 훨씬 더 성능이 좋은 것을 확인할 수 있다. 따라서 두 모델 중 가장 최적 성능을 보이는 모델은 랜덤 포레스트이다.